enhance/tools-dummy-data-react

Aunshon opened the pull request:
### All Submissions:

* [x] My code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)
* [x] My code satisfies feature requirements
* [x] My code is tested
* [x] My code passes the PHPCS tests
* [x] My code has proper inline documentation
* [x] I've included related pull request(s) (optional)
* [x] I've included developer documentation (optional)
* [x] I've added proper labels to this pull request

### Related Pull Request(s)

* https://github.com/getdokan/dokan-pro/pull/5177


### Closes 

* Closes https://github.com/getdokan/dokan-pro/issues/4680


### How to test the changes in this Pull Request:
* Test tools and dummy data page


### Changelog entry
```txt
- **update:** Admin Dashboard Dummy data importer page UI with Modern Design and Improved User Experience.
```

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Release Notes

* **New Features**
  * Added a comprehensive dummy data importer tool to the admin dashboard
  * Enables bulk import of vendors and products with real-time progress tracking
  * Provides completion status view with actions to review imported vendors and products
  * Includes data clearing functionality to reset and re-import dummy data
  * Extended admin navigation with new "Tools" and "Data Importer" sections
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Aunshon committed:
feat: add dummy-data route and component to admin dashboard

Aunshon committed:
feat: implement dummy-data components for admin dashboard

Aunshon committed:
feat: enhance dummy-data module with import, status, and clear functionality

Aunshon committed:
feat: improve dummy-data module with confirmation modal, error toast, and enhanced navigation

Aunshon committed:
feat: enhance dummy-data module with success toasts and loading skeleton

Aunshon committed:
fix: update dummy-data URL to use admin.php for accurate navigation

Aunshon committed:
fix: update dummy-data components with consistent syntax and styling adjustments

Aunshon committed:
fix: correct typo in dummy-data importer description text

Aunshon committed:
feat: add TypeScript types for Result component props

Aunshon committed:
fix: replace role and tabIndex with itemType for button accessibility in Result component

Aunshon committed:
fix: remove unused dummyData state and enforce type safety in dummy-data module

Aunshon committed:
feat: refactor dummy data handling by moving removable data constants to a separate module

Aunshon committed:
feat: simplify importer body rendering logic in index.tsx

Commit:
Merge branch 'develop' into enhance/tools-dummy-data-react

Resolve merge conflicts in Dashboard.php, Dashboard.tsx, and PanelSwitch.tsx
to integrate develop branch changes with dummy-data React feature.

MdAsifHossainNadim commented on src/admin/dashboard/pages/dummy-data/removableFields.ts:
```diff
@@ -0,0 +1,82 @@
+    'sku',
+    'status',
+    'catalog_visibility',
+    'short_description',
```
### [CRITICAL] Swapped export names

It looks like the export names got mixed up here. `vendorsDataToRemove` currently holds product-related fields (`sku`, `stock_status`, `sale_price`, `category_ids`, etc.), while `productsDataToRemove` holds vendor-related fields (`email`, `password`, `store_name`, `payment`, etc.).

When `formatVendorData()` calls `vendorsDataToRemove.forEach(key => delete data[key])`, it strips product fields from vendor data instead of vendor fields — and vice versa for products. This would result in malformed data being sent to the import endpoint.

Also, `status` appears twice in this array (lines 3 and 37).

The file name could also be improved — "Datas" isn't standard English, and "Obj" is misleading since these are arrays. Something like `removableFields.ts` would be clearer.

**Suggested fix — swap the names and remove the duplicate:**

```ts
// removableFields.ts

// Fields to strip from vendor CSV rows before import
export const vendorsDataToRemove = [
    'email',
    'password',
    'store_name',
    'social',
    'payment',
    'phone',
    // ... (the current productsDataToRemove array contents)
];

// Fields to strip from product CSV rows before import
export const productsDataToRemove = [
    'sku',
    'status',  // only once
    'catalog_visibility',
    'short_description',
    // ... (the current vendorsDataToRemove array contents, without duplicate 'status')
];
```

MdAsifHossainNadim commented on src/admin/dashboard/pages/dummy-data/index.tsx:
```diff
@@ -0,0 +1,312 @@
+                    ),
+                } );
+            } );
+    }
```
### [CRITICAL] Success toast fires on error

The success toast is inside `.finally()`, which runs regardless of whether the promise resolved or rejected. So when clearing fails, the user sees an error toast from `.catch()` followed immediately by a success toast from `.finally()` — which is confusing.

Also a grammar fix: `'remove'` should be `'removed'`.

**Suggested fix — move the success toast into `.then()`:**

```ts
function clearAllDummyData( setIsConfirmOpen: ( value: boolean ) => void ) {
    setLoading( true );
    apiFetch( {
        path: 'dokan/v1/dummy-data/clear',
        method: 'DELETE',
    } )
        .then( () => {
            resetToImport();
            loadImportStatus();
            loadCsvFile();
            toast( {
                type: 'success',
                title: __( 'All dummy data removed successfully', 'dokan-lite' ),
            } );
        } )
        .catch( ( err: any ) => {
            toast( {
                type: 'error',
                title: err?.message || __( 'Something went wrong', 'dokan-lite' ),
            } );
        } )
        .finally( () => {
            setLoading( false );
            setIsConfirmOpen( false );
        } );
}
```

MdAsifHossainNadim commented on src/admin/dashboard/components/Dashboard.tsx:
```diff
@@ -14,6 +14,7 @@ import VendorsPage from '../pages/vendors';
+import DummyData from "admin/dashboard/pages/dummy-data";
 
 export type DokanAdminRoute = {
     id: string;
```
### [CRITICAL] Route belongs in Pro, not Lite

This `dummy-data` route is a sub-page of Pro's Tools page — it even has a "Back to Tools" link that only works when Pro is active. The codebase already provides `wp.hooks.applyFilters('dokan-admin-dashboard-routes', routes)` on line 84-88 specifically so Pro can inject its own routes without modifying lite.

**Suggested fix — remove from here and register from Pro side:**

```tsx
// In Pro's dashboard index.tsx — register via the filter:
wp.hooks.addFilter( 'dokan-admin-dashboard-routes', 'dokan-pro', ( routes ) => [
    ...routes,
    { id: 'tools', element: <Tools />, path: '/tools' },
    { id: 'dummy-data', element: <DummyData />, path: '/dummy-data' },
] );
```

Also, the import uses double quotes while every other import in this file uses single quotes (WordPress/ESLint convention).

MdAsifHossainNadim commented on src/admin/dashboard/pages/dummy-data/Result.tsx:
```diff
@@ -0,0 +1,101 @@
+            { /* eslint-disable-next-line jsx-a11y/click-events-have-key-events,jsx-a11y/no-static-element-interactions */ }
+            <div
+                className="mb-[14px] border border-[#E9E9E9] rounded-[6px] flex items-center gap-[12px] p-[12px] hover:bg-[#F8F9F8]"
+                itemType="button"
```
### Accessibility — clickable divs with incorrect semantics

This `<div>` has an `onClick` handler with eslint a11y rules suppressed. `itemType="button"` is actually a microdata/schema.org attribute, not an ARIA role, so it doesn't help with keyboard navigation or screen readers. Users who rely on keyboard or assistive technology won't be able to interact with these.

Same issue on the "View Dummy Products" div below.

**Suggested fix — use semantic elements:**

```tsx
<button
    type="button"
    className="w-full mb-[14px] border border-[#E9E9E9] rounded-[6px] flex items-center gap-[12px] p-[12px] hover:bg-[#F8F9F8] bg-transparent cursor-pointer"
    onClick={ () => navigate( '/vendors' ) }
>
    <div className="h-[44px] w-[44px] border border-[#E9E9E9] bg-[#F8F9F8] flex justify-center items-center rounded-[5px]">
        <Users size="24" color="#828282" />
    </div>
    <p className="text-[14px] font-[600] text-[#25252D]">
        { __( 'View Dummy Vendors', 'dokan-lite' ) }
    </p>
</button>
```

For "View Dummy Products", use an `<a>` tag since it navigates to a different WP page.

MdAsifHossainNadim commented on src/admin/dashboard/pages/dummy-data/index.tsx:
```diff
@@ -0,0 +1,312 @@
+        }
+
+        const data = {
+            nonce: ( window as any )?.dokanAdminDashboard?.nonce,
```
### Nonce in REST body is unnecessary

`@wordpress/api-fetch` automatically attaches the nonce as an `X-WP-Nonce` header on every request — that's how WordPress REST authentication works. The nonce being sent inside the request body isn't consumed by the `DummyDataController` on the server side, so it doesn't add any security. It does unnecessarily expose the nonce token in request payloads though.

Same issue in `clearAllDummyData()` below.

**Suggested fix — remove the nonce from the data objects:**

```ts
const data = {
    vendor_data: vendorData,
    vendor_products: getVendorProducts( vendorData.id ),
    vendor_index: vendorIndex,
    total_vendors: allVendors.length,
};
```

MdAsifHossainNadim commented on src/admin/dashboard/pages/dummy-data/index.tsx:
```diff
@@ -0,0 +1,312 @@
+function Index( props: RouterProps ) {
+    const toast = useToast();
+    const [ progress, setProgress ] = useState( 0 );
+    const [ loading, setLoading ] = useState< boolean >( true );
```
### TypeScript `any` types should be replaced with proper interfaces

The project has TypeScript strict mode enabled, but this file uses `any` extensively — `useState<any[]>`, function params as `any`, API responses as `any`. Since the CSV structure is well-known (defined by the removable fields arrays), proper types are feasible and would catch bugs at compile time.

**Suggested approach:**

```ts
interface VendorRow {
    id: string;
    type: 'vendor';
    first_name: string;
    last_name: string;
    username: string;
}

interface ProductRow {
    type: string;
    name: string;
    description: string;
    vendor: string;
    raw_attributes?: RawAttribute[];
    manage_stock?: boolean;
}

interface ImportResponse {
    vendor_index: number;
}

const [ allVendors, setAllVendors ] = useState< VendorRow[] >( [] );
const [ allProducts, setAllProducts ] = useState< ProductRow[] >( [] );
```

MdAsifHossainNadim commented on src/admin/dashboard/pages/dummy-data/index.tsx:
```diff
@@ -0,0 +1,312 @@
+                    // @ts-ignore
+                    `${ dokanAdminDashboard.urls.adminRoot }admin.php?page=dokan-dashboard#/tools`
+                }
+                className="flex flex-row w-auto items-center gap-1 !text-[#828282] font-[400] text-[14px] hover:!underline mb-[20px]"
```
### "Back to Tools" causes full page reload

This constructs a full URL with `admin.php?page=dokan-dashboard#/tools`, which triggers a full browser navigation. Since both `/tools` and `/dummy-data` live in the same React hash router, this reload is unnecessary and breaks the SPA experience (loses in-memory state, flashes the page).

**Suggested fix — use React Router navigation:**

```tsx
import { useNavigate } from 'react-router-dom';

const navigate = useNavigate();

<button
    type="button"
    onClick={ () => navigate( '/tools' ) }
    className="flex flex-row w-auto items-center gap-1 text-[#828282] font-[400] text-[14px] hover:underline mb-[20px] bg-transparent border-none cursor-pointer"
>
    <ChevronLeft size="15" />
    <span>{ __( 'Back to Tools', 'dokan-lite' ) }</span>
</button>
```

MdAsifHossainNadim commented on src/admin/dashboard/pages/dummy-data/index.tsx:
```diff
@@ -0,0 +1,312 @@
+        setProgress( ( 100 * numVendorSucceed ) / allVendors.length );
+    }
+
+    function handleImport( vendorIndex: number ) {
```
### No cleanup on unmount during import

`handleImport()` is a recursive function that chains API calls one after another. If the user navigates to a different page while an import is in progress, the chain keeps firing and tries to call `setState` on an unmounted component. This can cause React warnings and unexpected behavior.

**Suggested fix — use an AbortController and clean up in useEffect:**

```tsx
useEffect( () => {
    const controller = new AbortController();
    abortRef.current = controller;

    loadImportStatus();
    loadCsvFile();

    return () => {
        controller.abort();
    };
}, [] );

function requestToImport( data: any ) {
    return apiFetch( {
        path: 'dokan/v1/dummy-data/import',
        method: 'POST',
        data,
        signal: abortRef.current?.signal,
    } );
}
```

MdAsifHossainNadim commented on src/admin/dashboard/pages/dummy-data/Result.tsx:
```diff
@@ -0,0 +1,101 @@
+                confirmButtonVariant="danger"
+                confirmationTitle={ __( 'Confirmation', 'dokan-lite' ) }
+                confirmationDescription={ __(
+                    'Are you sure to remove all dummy data!',
```
### Grammar in confirmation dialog

Small grammar fix:

```ts
// Current:
__( 'Are you sure to remove all dummy data!', 'dokan-lite' )

// Suggested:
__( 'Are you sure you want to remove all dummy data?', 'dokan-lite' )
```

MdAsifHossainNadim reviewed (changes requested):
## PR #3002 Review: enhance/tools-dummy-data-react

**Files Changed:** 8 (all frontend, +687/-0)

---

### Summary

Adds a React-based Dummy Data Importer page to the admin dashboard. Parses a CSV file client-side via `papaparse`, then sequentially POSTs each vendor + products to the existing `dokan/v1/dummy-data/import` REST endpoint. The related Pro PR (#5177) adds a Tools page, `ToolsController` REST API, and `ToolsActions` service.

Overall this is a solid feature — the component decomposition is clean and the UX flow makes sense. I do have some feedback items that need attention before merging. Please see the inline comments on the file changes and the items below.

---

### Pro PR (#5177) Feedback

These should be addressed in the related PR:

**11. [CRITICAL] `ToolsController` extends `WP_REST_Controller` directly** — Should extend `DokanRESTAdminController` or `DokanRESTController` for standard response formatting, HATEOAS links, and extensibility filters.

**12. Mixed indentation in `Assets.php`** — Likely a merge artifact. Run `composer phpcbf -- includes/Assets.php`.

**13. Route naming: `create_pages` uses underscore** — All other routes use kebab-case. Should be `create-pages`.

**14. Read-only endpoint uses POST** — `check-all-dokan-pages-exists` only reads data. Should use `WP_REST_Server::READABLE`.

**15. `ToolsAdminMenu` directly instantiated** — Should go through the DI container / service provider per project convention.

---

### PR Checklist

| Item | Status | Notes |
|---|---|---|
| Code satisfies requirements | Needs fix | Swapped field names is a correctness bug |
| Coding standards | Needs fix | Quote style, `any` types, eslint-disable usage |
| Screenshots | Missing | Required for UI changes |
| Tests | Missing | No PHPUnit or Playwright coverage for new page |
| Changelog entry | Pass | Present |
| Labels | Pass | Suggest adding: `Dependency With Pro` |

Aunshon committed:
refactor: restructure dummy-data module for improved maintainability

- Removed `RemovableDatasObj.ts` and migrated constants to a new `removableFields.ts` module.
- Enhanced component typing and replaced repetitive `any` usage with specific TypeScript types.
- Updated navigation logic and replaced deprecated elements for better accessibility.
- Improved error handling with consistent toast messages and loading states.

Aunshon committed:
Merge branch 'develop' into enhance/tools-dummy-data-react

Aunshon committed:
feat: add dummy-data route and reorganize removable fields for vendors and products

Aunshon committed:
fix: update dashboard URL for dummy data import link

MdAsifHossainNadim committed:
Merge branch 'develop' into enhance/tools-dummy-data-react
